File: maxProductSubarray.py
def maxProductSubArray(arr, i =0):
    N = len(arr)
    if i == N -2:
        return max(arr[0], arr[1], arr[0]*arr[1])
    if i == N - 1:
        return arr[0]
    
    
    
    res = maxProductSubArray(arr, i+1)

    return max(res, res*arr[i], arr[i])


print(maxProductSubArray([6 , -3, -10, 0 , 2 ]))
print(maxProductSubArray([-8, -6, -1]))
print(maxProductSubArray([2, 3, 4, 5, -1, 0]))

File: maxProductSubarray.js
function maxProduct(arr) {
	let max_e = arr[0];
	let min_e = arr[0];
	let global_max = arr[0];

	for (let i = 1; i < arr.length; i++) {
		let num = arr[i];
		const temp = max_e;

		max_e = Math.max(num, num * max_e, num * min_e);
		min_e = Math.min(num, num * temp, num * min_e);

		global_max = Math.max(global_max, max_e);
	}

	return BigInt(global_max);
}

console.log(
	maxProduct(
		'3 12 15 23 33 -35 30 -40 -30 -30 -30 26 28'.split(' ').map((x) => +x)
	)
);
